### PR TITLE
Build darwin/aarch64 and universal binary 

### PR DESCRIPTION
**Pull request by AndersonQ**
## What does this PR do?

It enables the Elastic-Agent and Beats to be built for `darwin/arm64` and generates a universal binary (`darwin/universal`).

The [universal binary](https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary) is the merge of the `darwin/arm64` with `darwin/amd64`. Therefore it requires to first build the  `darwin/arm64` and `darwin/amd64` and them _"merge"_ them together into the `darwin/universal`. Also `darwin/universal` isn't a valid platform or OS/ARCH pair, therefore it isn't recognised by the Go compiler. Given that, the choice was to add `darwin/arm64` as a new platform for the Elastic-Agent and when ever both platforms (`darwin/arm64` and `darwin/amd64`) are built, also generate and package the universal binary.
Even though  it's ~quite~ an implicit behaviour, the alternative would also require a lot of changes and exceptions to have our build process recognising `darwin/universal`, but compiling `darwin/arm64` and `darwin/amd64` and merging them. Also the current process makes available all three of them, allowing the user to choose which fits better their needs. The drawback of the universal binary is its size, the double of the single binaries.

A new target, `BuildDarwinUniversal`  was added on the `x-pack`s `magefile`s required by the Elastic-Agent.

Also fixes some small typos, redundant information on the help command and remove the use of some deprecated functions.

## Why is it important?

We need to support Mac with M1 chips

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [x] build darwin/arm64 binary
- [x] build universal 2 binary
- [x] use mage to automatically build the universal binary
- [ ] don't forget beats PR: https://github.com/elastic/beats/issues/29582

## How to test this PR locally
1. Set up the [beats repo](https://github.com/elastic/beats/):
   - have the beats repo on the same level as the `elastic-agent` repo
   - set the beats repo to the branch `26683-agent-arm64` while https://github.com/elastic/beats/pull/29585 isn't merged
1. Build the elastic-agent for the new platforms:
   - **darwin/arm64**: `PLATFORMS="darwin/arm64" PACKAGES="tar.gz" mage package`
   - **darwin/universal**: `PLATFORMS="darwin/arm64,darwin/amd64" PACKAGES="tar.gz" mage package`

3. Run the elastic-agent

####**Watch out:**
The elastic-agent tries to download any binary it might need to comply with the assigned policy. However none of the new binaries from this PR are present on the artifact API and other applications (e.g. fleet-server and elastic-endpoint) also aren't publishing `darwin/arm64` versions. Therefore if the elastic-agent tries to download them, it'll fail. This is the expected behavior and this PR does not address it, see *Out of Scope*.

## Out of scope:
 - Handle downloading binaries that does not publish `darwin-universal` artifacts on the artifact API
 - Handle upgrading a `darwin/amd64` agent running under Rosetta2 on a Mac with M1 chip. This will be addressed on another PR.

## Related issues

- Relates elastic/elastic-agent#151
- Closes elastic/elastic-agent#204

### CODE REVIEW

**Review by blakerouse — APPROVED** [review]
Overall I think this looks good.

**Review by ph — COMMENTED** [review]
Change LGTM, thanks for improving the comments too!
I am currently testing locally 

Using this PR and the beats PR and the following command.

```
 EXTERNAL=false PLATFORMS="darwin/arm64" PACKAGES="tar.gz" mage package
```

**Comment by ph** [comment]
@AndersonQ Are we missing an update for osquerybeat in the beats repository PR? Running the above command yield the following error or maybe there is something missing in the osquerybeat build @aleksmaus  ?

```
>> package: Building osquerybeat type=tar.gz for platform=darwin/arm64
package ran for 58.257699773s
Error: failed building osquerybeat type=tar.gz for platform=darwin/arm64: failed adding file={Source:build/data/install/darwin/amd64/osquery-5.0.1.pkg Content: Template: Target:osquery-5.0.1.pkg Mode:-rw-r--r-- Config:false Modules:false Dep:<nil> Owner: SkipOnMissing:false Symlink:false} to tar: lstat build/data/install/darwin/amd64/osquery-5.0.1.pkg: no such file or directory
package ran for 4m56.647351202s
Error: exit status 1

**Comment by aleksmaus** [comment]
@AndersonQ 
running this in the beats/x-pack/osquerybeat
```
SNAPSHOT=true DEV=true PLATFORMS="darwin/arm64" mage -v package
```
leads to this error:
```
Running dependency: CrossBuildGoDaemon
Running dependency: CrossBuild
Skipping cross-build of target=buildGoDaemon because platforms list is empty.
exec: docker info -f {{ json .}}
Running dependency: github.com/elastic/beats/v7/dev-tools/mage.CrossBuild.func1
exec: go mod download
Running dependency: github.com/elastic/beats/v7/dev-tools/mage.buildMage
exec: mage -f -goos=linux -goarch=amd64 -compile build/mage-linux-amd64
crossBuild: Platform list = [{darwin/arm64 cgo}]
package ran for 5.708173097s
Error: unsupported cross build platform darwin/arm64
```

what is the expectation of the beat repo? anybody is working on adjusting it to run for darwin/arm64?

**Comment by AndersonQ** [comment]
I'm on it. It seems the removal of the elastic-agent code from the beats repo broke the M1 PR.

**Comment by aleksmaus** [comment]
> @aleksmaus @ph The elastic-agent needs to build the beats as well, as the `darwin-aarch64` isn't yet available on the artefact API. To do so the following is required:

> 
> I updated the PR description with these instructions

Thank you! Will have PR for osquerybeat shortly to support darwin/arm64.

File: dev-tools/mage/common.go [hunk 1]
```diff
@@ -573,7 +594,7 @@ func ParallelCtx(ctx context.Context, fns ...interface{}) {
 // Parallel runs the given functions in parallel with an upper limit set based
 // on GOMAXPROCS.
 func Parallel(fns ...interface{}) {
-	ParallelCtx(context.Background(), fns...)
+	ParallelCtx(context.TODO(), fns...)
```
Comment (ph) [inline dev-tools/mage/common.go:597]: 👍🏽

**Review by ph — APPROVED** [review]
LGTM, I've tested the build, I haven't tested on m1.

**Comment by aleksmaus** [comment]
@ph @AndersonQ it's still a problem with building on M1

```
Status: Downloaded newer image for docker.elastic.co/beats-dev/golang-crossbuild:1.17.8-darwin-arm64-debian10
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
>> Building using: cmd='build/mage-linux-arm64 golangCrossBuild', env=[CC=oa64-clang, CXX=oa64-clang++, GOARCH=arm64, GOARM=, GOOS=darwin, PLATFORM_ID=darwin-arm64]
# runtime/cgo
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
```

I already mentioned before that the docker image tagged with arm64 is actually amd64 arch and qemu can't run that

**Comment by amitkanfer** [comment]
@AndersonQ ,
The CSP team all have M1 laptops, and now that our cloudbeat is part of agent, it'll be great to be able to compile locally.
Do we have an ETA for the M1 issue to be resolved?
In the meantime. can we spin up a compilation server + script that will rsync the files from the local laptop, compile, and bring back the artifacts?

Thank you!

**Pull request merged**